src/Processors/Formats/Impl/Parquet/ParquetRecordReader.cpp

```diff
@@ -0,0 +1,237 @@
+#include "ParquetRecordReader.h"
+
+#include <bit>
+
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnsNumber.h>
+#include <Common/logger_useful.h>
+#include <Common/Stopwatch.h>
+#include <Core/Types.h>
+#include <DataTypes/DataTypeDate32.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/castColumn.h>
+
+#include <arrow/status.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/column_reader.h>
+#include <parquet/properties.h>
+
+#include "ParquetLeafColReader.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int PARQUET_EXCEPTION;
+}
+
+#define THROW_PARQUET_EXCEPTION(s)                                            \
+    do                                                                        \
+    {                                                                         \
+        try { (s); }                                                          \
+        catch (const ::parquet::ParquetException & e)                         \
+        {                                                                     \
+            auto msg = PreformattedMessage::create("Excepted when reading parquet: {}", e.what()); \
+            throw Exception(std::move(msg), ErrorCodes::PARQUET_EXCEPTION);   \
+        }                                                                     \
+    } while (false)
+
+namespace
+{
+
+std::unique_ptr<parquet::ParquetFileReader> createFileReader(
+    std::shared_ptr<::arrow::io::RandomAccessFile> arrow_file)
+{
+    std::unique_ptr<parquet::ParquetFileReader> res;
+    THROW_PARQUET_EXCEPTION(res = parquet::ParquetFileReader::Open(std::move(arrow_file)));
```

Comment:
File metadata should be passed in here, otherwise file footer will be read again here.

Reply:
OK